xds/src/main/java/io/grpc/xds/ClusterImplLoadBalancer.java

```diff
@@ -0,0 +1,283 @@
+ * the priority_experimental LB policy and the parent LB policy of the weighted_target_experimental
+ * LB policy in the xDS load balancing hierarchy.
+ */
+final class ClusterImplLoadBalancer extends LoadBalancer {
+
+  @VisibleForTesting
+  static final long DEFAULT_PER_CLUSTER_MAX_CONCURRENT_REQUESTS = 1024L;
+  @VisibleForTesting
+  static boolean enableCircuitBreaking =
+      Boolean.parseBoolean(System.getenv("GRPC_XDS_EXPERIMENTAL_CIRCUIT_BREAKING"));
+
+  private final XdsLogger logger;
+  private final Helper helper;
+  private final ThreadSafeRandom random;
+  // The following fields are effectively final.
+  private String cluster;
+  @Nullable
+  private String edsServiceName;
+  private ObjectPool<XdsClient> xdsClientPool;
+  private XdsClient xdsClient;
+  private CallCounterProvider callCounterProvider;
+  private LoadStatsStore loadStatsStore;
+  private RequestLimitingLbHelper childLbHelper;
+  private LoadBalancer childLb;
+
+  ClusterImplLoadBalancer(Helper helper) {
+    this(helper, ThreadSafeRandomImpl.instance);
+  }
+
+  ClusterImplLoadBalancer(Helper helper, ThreadSafeRandom random) {
+    this.helper = checkNotNull(helper, "helper");
+    this.random = checkNotNull(random, "random");
+    InternalLogId logId = InternalLogId.allocate("cluster-impl-lb", helper.getAuthority());
+    logger = XdsLogger.withLogId(logId);
+    logger.log(XdsLogLevel.INFO, "Created");
+  }
+
+  @Override
+  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    logger.log(XdsLogLevel.DEBUG, "Received resolution result: {0}", resolvedAddresses);
+    Attributes attributes = resolvedAddresses.getAttributes();
+    if (xdsClientPool == null) {
+      xdsClientPool = attributes.get(XdsAttributes.XDS_CLIENT_POOL);
+      xdsClient = xdsClientPool.getObject();
+    }
+    if (callCounterProvider == null) {
+      callCounterProvider = attributes.get(XdsAttributes.CALL_COUNTER_PROVIDER);
+    }
+    ClusterImplConfig config =
+        (ClusterImplConfig) resolvedAddresses.getLoadBalancingPolicyConfig();
+    if (cluster == null) {
+      cluster = config.cluster;
+      edsServiceName = config.edsServiceName;
+      childLbHelper = new RequestLimitingLbHelper(
+          callCounterProvider.getOrCreate(config.cluster, config.edsServiceName));
+      childLb = config.childPolicy.getProvider().newLoadBalancer(childLbHelper);
+      // Assume load report server does not change throughout cluster lifetime.
+      if (config.lrsServerName != null) {
+        if (config.lrsServerName.isEmpty()) {
+          loadStatsStore = xdsClient.addClientStats(cluster, edsServiceName);
+        } else {
+          logger.log(XdsLogLevel.WARNING, "Can only report load to the same management server");
+        }
+      }
+    }
+    childLbHelper.updateDropPolicies(config.dropCategories);
+    if (config.maxConcurrentRequests != null) {
+      childLbHelper.updateMaxConcurrentRequests(config.maxConcurrentRequests);
+    }
```

Comment:
Is that possible `maxConcurrentRequests` changes from non-null to null?

Reply:
Yeah, the code should be able to handle such updates. Fixed.